sdk/eventhub/azure-eventhubs/azure/eventhub/_consumer_producer_mixin.py

```diff
@@ -7,12 +7,30 @@
 import logging
 import time
 
-from uamqp import errors
+from uamqp import errors, constants, compat
 from azure.eventhub.error import EventHubError, _handle_exception
 
 log = logging.getLogger(__name__)
 
 
+def _retry_decorator(to_be_wrapped_func):
+    def wrapped_func(self, *args, **kwargs):
+        timeout = kwargs.pop("timeout", None)
+        if not timeout:
+            timeout = 100000  # timeout None or 0 mean no timeout. 100000 seconds is equivalent to no timeout
+        timeout_time = time.time() + timeout
+        max_retries = self.client.config.max_retries
+        retry_count = 0
+        last_exception = None
+        while True:
+            try:
+                return to_be_wrapped_func(self, timeout_time=timeout_time, last_exception=last_exception, **kwargs)
```

Comment:
This seems like it will blow up if the caller supplies `timeout_time` on their own as a keyword arg when they call the function, and will be silently ignored if they supply it on their own as a positional argument. I would say in general it's unusual for a decorator to supply more than `*args, **kwargs` to the called func. If this decorator is really a helper to supply arguments like `timeout_time` and `last_exception` to a function, and these args are really intended to be *non-public* (i.e. nothing but the decorator should supply them), then I would suggest:

* removing them as named parameters of the wrapped functions
* only passing them via kwargs
* prefixing the names with underscores `"_timeout_time"` and `"_last_exception"`

That's a little unusual, but makes it clear that no caller is supposed to provide these values (and all makes it less likely that anyone/anything will try to). 

Alternatively, it's possible there is some other design that does not require the wrapped function to accept these values directly itself. I would actually say that is worth investigating.